impl/src/main/java/com/okta/sdk/impl/util/OAuth2Utils.java

```diff
@@ -0,0 +1,163 @@
+
+    private static KeyFactory getKeyFactoryInstance() throws NoSuchAlgorithmException {
+        return KeyFactory.getInstance("RSA");
+    }
+
+    public static String createSignedJWT(ClientConfiguration clientConfiguration)
+        throws RuntimeException, InvalidKeySpecException, NoSuchAlgorithmException, IOException {
+        String clientId = clientConfiguration.getClientId();
+        SignatureAlgorithm signatureAlgorithm = getSignatureAlgorithm(clientConfiguration.getAlgorithm());
+        PrivateKey privateKey = createPrivateKeyFromPemFilePath(clientConfiguration.getPrivateKey());
+        Instant now = Instant.now();
+
+        String jwt = Jwts.builder()
+            .setAudience(clientConfiguration.getBaseUrl() + "/oauth2/default/v1/token")
+            .setIssuedAt(Date.from(now))
+            .setExpiration(Date.from(now.plus(1L, ChronoUnit.HOURS)))
+            .setIssuer(clientId)
+            .setSubject(clientId)
+            .claim("jti", UUID.randomUUID().toString())
+            .signWith(signatureAlgorithm, privateKey)
+            .compact();
+
+        return jwt;
+    }
+
+    public static String getOAuth2AccessToken(ClientConfiguration clientConfiguration)
+        throws HttpException, IOException, InvalidKeySpecException, NoSuchAlgorithmException {
+        String signedJwt = createSignedJWT(clientConfiguration);
+
+        RequestExecutor requestExecutor = okHttpRequestExecutorFactory.create(clientConfiguration);
+
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.add("Accept", "application/json");
+        httpHeaders.add("Content-Type", "application/x-www-form-urlencoded");
+
+        QueryString queryString = new QueryString();
+        queryString.put("grant_type", "client_credentials");
+        queryString.put("scope", clientConfiguration.getScopes());
+        queryString.put("client_assertion_type",
+            URLEncoder.encode("urn:ietf:params:oauth:client-assertion-type:jwt-bearer", "UTF-8"));
+        queryString.put("client_assertion", signedJwt);
+
+        DefaultRequest accessTokenRequest =
+            new DefaultRequest(HttpMethod.POST, "/oauth2/v1/token", queryString, httpHeaders);
+        Response accessTokenResponse = requestExecutor.executeRequest(accessTokenRequest);
+
+        if (accessTokenResponse != null && accessTokenResponse.getBody() != null) {
+            Reader reader = new InputStreamReader(accessTokenResponse.getBody(), StandardCharsets.UTF_8);
+            Type mapType = new TypeToken<Map<String, String>>(){}.getType();
+            Map<String, String> responseMap = gson.fromJson(reader, mapType);
+
+            if (accessTokenResponse.getHttpStatus() == HttpURLConnection.HTTP_OK) {
+                return responseMap.get("access_token");
+            }
+            else {
+                String error = responseMap.get("error");
+                String error_description = responseMap.get("error_description");
```

Comment:
Camels, not snakes ;)

Reply:
Response from Auth Server contains camel cased fields isn't it?

`{
    "token_type": "Bearer",
    "expires_in": 3600,
    "access_token": "eyJraWQiOiJ.....UfThlJ7w",
    "scope": "{scopes}"
}`

and in error cases:

`{
  "error": "invalid_client",
  "error_description": "{Error Details}"
}`

Reply:
yes, but you `String error_description` should be `String errorDescription`

Reply:
Got it now :)